ledger/src/blocktree.rs

```diff
@@ -3660,4 +3664,47 @@ pub mod tests {
             assert!(blocktree.get_slot_entries(slot, 0, None).is_err());
         }
     }
+
+    #[test]
+    fn test_multiple_threads_insert_shred() {
+        let blocktree_path = get_tmp_ledger_path!();
+        let blocktree = Arc::new(Blocktree::open(&blocktree_path).unwrap());
+
+        for _ in 0..100 {
```

Comment:
why 100? why 10 threads? This looks like a regression test, not a unit test. I'd expect to see `2` in a unit test - something minimal to showcase the problem being solved. How long does this test take to execute?

Reply:
This is to make the race condition more likely to repro, right now the test takes around 2-3 seconds. The failure case is essentially when thread 1 and thread 2 interleave read-writes, something like:

Read 1 (thread 1)
Read 2 (thread 2)
Write 1 (thread 1)
Write 2 (thread 2)

We could deterministically reproduce this interleaving by hacking in forced pauses into the critical section of the function itself for test purposes only, but in the past those types tests have been deleted due to upkeep/complexity they introduce.

Generally our approach to date for testing these types of race conditions is setting up situations in which they are likely to happen, but of course due to the nature of the bug, not 100% guaranteed.

How would you approach this?

Reply:
I'd put that expensive test into the integration test so that it doesn't slow down the `cargo watch -x 'test --lib'` workflow, and I'd add a bunch of comments saying how the test is less than ideal and that I don't know how to unit-test it. Then I'd add a GitHub issue saying that we don't know (or have time) to mock out these objects to write a proper unit test, and how it adds 2-3 seconds to each and everyone's test runs until we do.

Reply:
integration test suite == currently-nonexistent `tests` directory in `solana_ledger/`.